src/org/labkey/test/util/TestDataUtils.java

```diff
@@ -23,4 +24,60 @@ public static List<Map<String, Object>> rowMapsFromTsv(String tsvString) throws
             return new TabLoader.TsvFactory().createLoader(dataStream, true).load();
         }
     }
+
+    public static List<Map<String, Object>> rowMapsFromCsv(String tsvString) throws IOException
+    {
+        try (InputStream dataStream = IOUtils.toInputStream(tsvString, StandardCharsets.UTF_8))
+        {
+            return new TabLoader.CsvFactory().createLoader(dataStream, true).load();
+        }
+    }
+
+    public static String tsvStringFromRowMaps(List<Map<String, Object>> rowMaps, List<String> columns,
+                                              boolean includeHeaders)
+    {
+        return toTabular(rowMaps, columns, "\t", includeHeaders);
+    }
+
+    public static String csvStringFromRowMaps(List<Map<String, Object>> rowMaps, List<String> columns,
+                                              boolean includeHeaders)
+    {
+        return toTabular(rowMaps, columns, ",", includeHeaders);
+    }
+
+    /**
+     * Convert a list of Map<String, Object>> to tabluar (tsv, csv) format
+     * (assumes the rowMaps all share the same keyset/schema)
+     * can be used to generate edit-grid paste data, if delimiter is \t and includeHeaders is false
+     *
+     * @param rowMaps data to be written into tabular format
+     * @param columns the fields (in order) from the rowMaps to include in tabular output
+     * @param delimiter comma [,] for csv tab [\t] for tsv
+     * @param includeHeaders    whether to write the keys as column names on the first line of the output string
+     * @return
+     */
+    private static String toTabular(List<Map<String, Object>> rowMaps, List<String> columns,
+                                    String delimiter, boolean includeHeaders)
+    {
+        StringBuilder builder = new StringBuilder();
+
+        if (includeHeaders)
+        {
+            builder.append(String.join(delimiter, columns));
+            builder.append("\n");
+        }
+
+        for (Map<String, Object> row : rowMaps)
+        {
+            List<String> values = new ArrayList<>();
+            for (String name : columns)
+            {
+                Object value = row.get(name);
+                values.add(value != null ? String.valueOf(value) : "");
+            }
+            builder.append(String.join(delimiter, values));
+            builder.append("\n");
+        }
+        return builder.toString();
```

Comment:
Consider using the `TsvMapWriter` from the server API. I don't love referencing stuff in the server API but the dependency is already there so no harm in using it.
```suggestion
        StringBuilder builder = new StringBuilder();
        try (TSVWriter writer = new TSVMapWriter(rows))
        {
            writer.setHeaderRowVisible(includeHeaders);
            writer.setDelimiterCharacter(delimiter);
            writer.write(builder);
        }
        return builder.toString();
```
Check `org.labkey.api.data.TSVMapWriter.Tests` for more examples.

Reply:
When I try using `TsvMapWriter`, it builds fine but at runtime I am shown this error:
```

C:\dev\labkey\develop\server\testAutomation\src\org\labkey\test\util\TestDataUtils.java:114: error: cannot access HttpServletResponse
            writer.write(builder);
                  ^
  class file for jakarta.servlet.http.HttpServletResponse not found
``` 🙍

Reply:
Ok. I guess there IS some harm in using the server api 😛